### PR TITLE
chore: add pbandk protobuf (AR-1303)

### PR DESCRIPTION
**Pull request by vitorhugods**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] contains a reference JIRA issue number like `SQPIT-764`
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

We need a multiplatform protobuf solution that covers all our usecases

### Solutions

Add [PBandK](https://github.com/streem/pbandk).

How it is setup:
Two new modules:
* `protobuf-codegen` which is a "Java Library" that includes the `.proto` files. It will call Google's Gradle Protobuf plugin to generate the code, with the PBandK add-on.
* `protobuf` which is a bit of a hack, as Google's Protobuf plugin doesn't work with Kotlin Multiplatform yet. This module depends on `protobuf-codegen` and will trigger a code generation there, adding the generated files to the `commonMain` SourceSet. This module also contains the PBandK in `commonMain`, which allows all targets to encode and decode to/from byte arrays.

The idea was taken from [a sample in pbandk's repo](https://github.com/streem/pbandk/tree/master/examples/browser-js). As [suggested in this issue/discussion](https://github.com/streem/pbandk/issues/33).

### Testing

Add a simple test to serve as a sample and sanity check.

### Notes

- This current implementation replaces the current Protobuf utilization to encode/decode messages, but doesn't replace the creation of envelopes yet. This should be trivial and tackled in a next PR to make this one smaller.
- The 475+ lines of `.proto` files are copy/pasted from our [generic-message-proto](https://github.com/wireapp/generic-message-proto/tree/master/proto) repo.
- Both `messages.proto` and `otr.proto` contained a `QualifiedUserID`. So I've extracted it to its own dedicated file called `qualified_user_id.proto`.
- Naming is an issue and we probably want to solve this before merging. Especially for `ClientId` and `UserId`, which can be misleading, making discoverability through autocompletion a bit annoying.

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapper.kt [hunk 1]
```diff
@@ -1,9 +1,66 @@
...
+            is GenericMessage.Content.Image -> MessageContent.Unknown
+            is GenericMessage.Content.Knock -> MessageContent.Unknown
+            is GenericMessage.Content.LastRead -> MessageContent.Unknown
+            is GenericMessage.Content.Location -> MessageContent.Unknown
+            is GenericMessage.Content.Reaction -> MessageContent.Unknown
+            null -> {
+                kaliumLogger.w("Null content when parsing protobuf. Message UUID = $genericMessage.")
+                MessageContent.Unknown
+            }
+        }
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapper.kt:62]: Exhaustive `when` for protobufs!

File: buildSrc/src/main/java/Dependencies.kt [hunk 2]
```diff
@@ -148,8 +154,11 @@ object Dependencies {
     }
 
     object Protobuf {
+        @Deprecated("A new implementation is avaliable. Use the protobuf project instead.")
```
Comment (typfel) [inline buildSrc/src/main/java/Dependencies.kt:157]: Why aren't we deleting this? The implementation has been deleted, no?

File: build.gradle.kts [hunk 3]
```diff
@@ -19,6 +19,7 @@ buildscript {
         classpath("app.cash.sqldelight:gradle-plugin:$sqlDelightVersion")
         classpath("com.wire:carthage-gradle-plugin:0.0.1")
         classpath("org.jetbrains.dokka:dokka-gradle-plugin:$dokkaVersion")
+        classpath("com.google.protobuf:protobuf-gradle-plugin:0.8.18")
```
Comment (alexandreferris) [inline build.gradle.kts:22]: variable on plugin version

File: buildSrc/src/main/java/Dependencies.kt [hunk 4]
```diff
@@ -148,8 +154,11 @@ object Dependencies {
...
         const val wireJvmMessageProto = "com.wire:generic-message-proto:${Versions.wireJvmMessageProto}"
+        @Deprecated("A new implementation is avaliable. Use the protobuf project instead.")
```
Comment (typfel) [inline buildSrc/src/main/java/Dependencies.kt:159]: Also delete?

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapperProvider.kt [hunk 5]
```diff
@@ -1,3 +1,2 @@
 package com.wire.kalium.logic.data.message
 
-expect fun provideProtoContentMapper(): ProtoContentMapper
```
Comment (alexandreferris) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapperProvider.kt:3]: Why do we still have this class/file if it is not being used anymore? Can't we delete it? 🤔

**Review by alexandreferris — COMMENTED** [review]
Great work!

I just have two comments 🤔

File: protobuf-codegen/src/main/proto/messages.proto [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+syntax = "proto2";
+
+import "qualified_user_id.proto";
+
+message GenericMessage {
```
Comment (typfel) [inline protobuf-codegen/src/main/proto/messages.proto:24]: The proto files are copied into the project now? This makes it hard to know which of the protobuf file we are using in Kalium. 

Can we pull them GitHub only define the release tag in the project. If this is not easy we could do it in a separate PR  but I think we should do it.

**Review by typfel — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/java/Dependencies.kt (hunk 2 above)
Comment (vitorhugods) [reply buildSrc/src/main/java/Dependencies.kt:157]: Still used in Network for envelope serialisation.

I want to do that in another PR.

File: buildSrc/src/main/java/Dependencies.kt (hunk 4 above)
Comment (vitorhugods) [reply buildSrc/src/main/java/Dependencies.kt:159]: Same as above. Still used for envelope creation.

File: protobuf-codegen/src/main/proto/messages.proto (hunk 6 above)
Comment (vitorhugods) [reply protobuf-codegen/src/main/proto/messages.proto:24]: Not 100% copy pasted. As I mentioned in the PR description, the QualifiedUserID, for example, needed to be extracted to a common file.

Some of the Java options supported by Protoc (like outer class name or adding a suffix) are not respected when using PBandK.

I'd like to streamline the process and checkout from the otr-proto repo, but it's far from easy to do so right now. Maybe it would help by extracting the QualifiedUserID in the original otr-proto repo, but that would be a breaking change.

I don't have any solutions now.

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapperProvider.kt (hunk 5 above)
Comment (vitorhugods) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapperProvider.kt:3]: Oops hahahah

I'll delete it tomorrow when I'm back.

File: build.gradle.kts (hunk 3 above)
Comment (vitorhugods) [reply build.gradle.kts:22]: It's inside the `buildScript` block, so we can't do it 😔

The `buildScript` block is VERY special.

**Review by typfel — APPROVED** [review]

File: protobuf-codegen/src/main/proto/messages.proto (hunk 6 above)
Comment (typfel) [reply protobuf-codegen/src/main/proto/messages.proto:24]: The `otr.proto` and `messages.proto` are supposed to live in different namesakes doesn't PBandK respect that? See https://github.com/wireapp/generic-message-proto/pull/74#pullrequestreview-680786841

Comment (typfel) [reply protobuf-codegen/src/main/proto/messages.proto:24]: So yeah, seems like https://github.com/streem/pbandk/issues/179 doesn't respect `package` but also the `option java_package` is different so I don't understand why the clash.

Comment (typfel) [reply protobuf-codegen/src/main/proto/messages.proto:24]: Right so it clashes because package name inside the proto files are override by the plugin option:
`option("kotlin_package=com.wire.kalium.protobuf")`

Comment (vitorhugods) [reply protobuf-codegen/src/main/proto/messages.proto:24]: TBH I felt like PBandK doesn't respect a lot of these options yet. I tried playing with them before submitting the PR and in the end I just removed them, as they didn't have any effect on the generated code.
I didn't want to leave `package` and `option`s in there, so they would suddenly kick-in when we bump PBandK's version or someone spends time trying to understand these lines of code in the future.

It's a trade-off for now. We increase a bit of maintenance with `.proto` definitions, but as a pro: we have a multi-platform solution. In my opinion, it's worth it. But I'd like a second confirmation before we move on.

Hopefuly these will be supported Soon™.

Comment (typfel) [reply protobuf-codegen/src/main/proto/messages.proto:24]: Yes, it's definitely worth it! Just looking for way to avoid having to copy & paste the files and edit them. But like I said earlier I'm not pushing for solving it in this PR.

Comment (vitorhugods) [reply protobuf-codegen/src/main/proto/messages.proto:24]: > Right so it clashes because package name inside the proto files are override by the plugin option: `option("kotlin_package=com.wire.kalium.protobuf")`

Indeed, the `kotlin_package` was overwriting it 🤦🏼‍♂️. Removing that made the `package` work.

Still, the original `messages.proto` doesn't have a `package`. So we still need to modify the files a tiny bit when getting from the original repo. But the changes are one-liner.

File: build.gradle.kts (hunk 3 above)
Comment (alexandreferris) [reply build.gradle.kts]: ayayay didn't see that, sorry 🙈

**Review by alexandreferris — APPROVED** [review]
